server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java

```diff
@@ -201,6 +201,22 @@ NumericType numericType() {
          */
         public abstract long roundUpToMillis(long value);
 
+        public final void validateTimestamp(long value, DocumentParserContext context) {
+            if (context.indexSettings().getTimeSeriesStartTime() == null) {
+                assert context.indexSettings().getTimeSeriesEndTime() == null;
+                return;
+            }
+            long startTime = convert(context.indexSettings().getTimeSeriesStartTime());
+            if (value < startTime) {
+                throw new IllegalArgumentException("time series index @timestamp value [" + value + "] must be larger than " + startTime);
+            }
+
+            long endTime = convert(context.indexSettings().getTimeSeriesStartTime());
+            if (value >= endTime) {
+                throw new IllegalArgumentException("time series index @timestamp value [" + value + "] must be smaller than " + endTime);
+            }
+        }
```

Comment:
I know I wrote this, but on further reading, I think maybe it makes sense to just move it all to `DataStreamTimestampFieldMapper`. I thought I was being clever, but all of this code should be able to run just fine from there and it'd be nice that `DateFieldMapper` doesn't need to care about the validation.

Reply:
I agree with moving this logic to `DataStreamTimestampFieldMapper`, since there there is already validation for the `@timestamp` field (whether it exists and whether a single value has been specified). And I think at least initially, tsdb will work only with data streams.